## ANA-522
## Mod03 Week05: Data Preparation 
## Due: Sunday at midnight

**Instructions:**

- Submit your work as a PDF or .ipynb file on blackboard.

- Please include your name in the filename of your submission

- Please make sure all code is executed and displays the corresponding output. If the corresponding output is not displayed, no points will be given.

- *Note:* Each question includes sample output that you can use to check if your solution is correct.

- *Additional Note:* Questions 12, 13, 17, 19, and 23 are optional exercises. You may choose to complete them for additional practice, but they are not required to finish in this assignment.

<br>

## Titanic Dataset
The sinking of the Titanic is one of the most infamous shipwrecks in history.  

On April 15, 1912, during her maiden voyage, the widely considered “unsinkable” RMS Titanic sank after colliding with an iceberg. Unfortunately, there weren’t enough lifeboats for everyone onboard, resulting in the death of 1502 out of 2224 passengers and crew.  

While there was some element of luck involved in surviving, it seems some groups of people were more likely to survive than others. With the availabilities of data analytic and machine learning technologies these days, people are researching in finding clues of survival from hidden data patterns.

Our purpose in the Lab exercise is to prepare and clean the dataset to be handy for data analysis tasks.

#### The Titanic dataset is  available on Blackboard:      


<img src="http://mathqs.mcdaniel.edu/ANA522/exam/Titanic_DD.png" width="800">

<img src="http://mathqs.mcdaniel.edu/ANA522/exam/Titanic_Var_Note.png" width="600">

In [1]:
import numpy as np
import pandas as pd

In [2]:
import os
print("Current working directory:", os.getcwd())
titanic = pd.read_csv('titanic.csv', sep=',')



Current working directory: /Users/nkohei/Workspace/McDaniel-Repository/522/week5


### Q01: Use a pandas function to overview the Titanic dataset.  
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_overview.png" 
     width="1000" height="600" alt="attributes list" />

In [3]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### Q02: Find a pandas attribute to display data type of each column ( all Titanic dataset properties. )  
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_att_types.png" 
     width="200" height="400" alt="attributes data types" />

In [4]:
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

### Q03: Display all entries whose Age attribute value is missing (NA).  
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_age_na.png" 
     width="1000" height="600" alt="attributes list" />

In [5]:
titanic[titanic['Age'].isna()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

### Q04: Show how many entries whose Embarked attribute value is missing (NA).

There are 2 entries whose Embarked values is NA

In [6]:
titanic[titanic['Embarked'].isna()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

### Q05: Display all entries whose Age attribute value or Cabin attribute value is missing (NA). 
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_age_or_cabin_na.png" 
     width="1000" height="600" alt="attributes list" />

In [7]:
titanic[titanic['Age'].isna() | titanic['Cabin'].isna()]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[706 rows x 12 columns]

### Q06: Show how many entries are there both Age and Cabin attribute values are missing (NA). 
There are 158 entries with NA in both Age and Cabin.

In [8]:
titanic[titanic['Age'].isna() & titanic['Cabin'].isna()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[158 rows x 12 columns]

### Q07: Sample any 15 entries to be sorted by PassengerId that both Age and Cabin attributes are missing (NA).
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_age_and_cabin_na_sample.png" 
     width="1000" height="600" alt="attributes list" />

In [9]:
titanic[titanic['Age'].isna() & titanic['Cabin'].isna()].sample(15)

PassengerId  Survived  Pclass  \
485          486         0       3   
563          564         0       3   
186          187         1       3   
250          251         0       3   
368          369         1       3   
384          385         0       3   
766          767         0       1   
274          275         1       3   
792          793         0       3   
375          376         1       1   
36            37         1       3   
598          599         0       3   
466          467         0       2   
126          127         0       3   
578          579         0       3   

                                                Name     Sex  Age  SibSp  \
485                           Lefebre, Miss. Jeannie  female  NaN      3   
563                                Simmons, Mr. John    male  NaN      0   
186  O'Brien, Mrs. Thomas (Johanna "Hannah" Godfrey)  female  NaN      1   
250                           Reed, Mr. James George    male  NaN      0   
368                              Jermyn, Miss. Annie  female  NaN      0   
384                           Plotcharsky, Mr. Vasil    male  NaN      0   
766                        Brewe, Dr. Arthur Jackson    male  NaN      0   
274                       Healy, Miss. Hanora "Nora"  female  NaN      0   
792                          Sage, Miss. Stella Anna  female  NaN      8   
375            Meyer, Mrs. Edgar Joseph (Leila Saks)  female  NaN      1   
36                                  Mamee, Mr. Hanna    male  NaN      0   
598                                Boulos, Mr. Hanna    male  NaN      0   
466                            Campbell, Mr. William    male  NaN      0   
126                              McMahon, Mr. Martin    male  NaN      0   
578                 Caram, Mrs. Joseph (Maria Elias)  female  NaN      1   

     Parch           Ticket     Fare Cabin Embarked  
485      1             4133  25.4667   NaN        S  
563      0  SOTON/OQ 392082   8.0500   NaN        S  
186      0           370365  15.5000   NaN        Q  
250      0           362316   7.2500   NaN        S  
368      0            14313   7.7500   NaN        Q  
384      0           349227   7.8958   NaN        S  
766      0           112379  39.6000   NaN        C  
274      0           370375   7.7500   NaN        Q  
792      2         CA. 2343  69.5500   NaN        S  
375      0         PC 17604  82.1708   NaN        C  
36       0             2677   7.2292   NaN        C  
598      0             2664   7.2250   NaN        C  
466      0           239853   0.0000   NaN        S  
126      0           370372   7.7500   NaN        Q  
578      0             2689  14.4583   NaN        C

### Q08: Display and show how many entries in the dataset have no missing value (NA) in all Age, Cabin, and Embarked columns.
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_no_na.png" 
     width="1000" height="600" alt="attributes list" />

In [10]:
# titanic[titanic['Age'].notna() & titanic['Cabin'].notna() & titanic['Embarked'].notna()]

titanic[~(titanic['Age'].isna() | titanic['Cabin'].isna() | titanic['Embarked'].isna())]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

### Q9: Create a new DataFrame, without modify the original, to hold all entries withtout any NA value of any attribute from the original Titanic dataset. 
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_no_na_df.png" 
     width="1000" height="600" alt="attributes list" />

In [11]:
titanic_no_na = titanic[~(titanic['Age'].isna() | titanic['Cabin'].isna() | titanic['Embarked'].isna())]

In [12]:
titanic_no_na

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

### Q10: Replace all missing values (NA) with 0 without overwriting the original dataset by createing/saving in a new DataFrame. Display the overview of the new dataset.
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_fill_na.png" 
     width="1000" height="600" alt="attributes list" />

In [13]:
titanic_fill_na = titanic.fillna(0)
titanic_fill_na

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   0.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500     0        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250     0        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500     0        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000     0        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500     0        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500     0        Q  

[891 rows x 12 columns]

### Q11: Replace all missing values (NA) in Age column with 0.0, and in Cabin with 'AXX' without overwriting the original dataset by createing/saving in a new DataFrame. Display the overview of the new dataset.
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_replace_na.png" 
     width="1000" height="600" alt="attributes list" />

In [14]:
titanic_replace_na = titanic.fillna({'Age': 0.0, 'Cabin': 'AXX'})
titanic_replace_na

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   0.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   AXX        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   AXX        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   AXX        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   AXX        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   AXX        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   AXX        Q  

[891 rows x 12 columns]

### *Optional*: Q12: Replace all missing values (NA) using forward filling method without overwriting the original dataset, by createing/saving in a new DataFrame. Display the overview of the new dataset.
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_ffill.png" 
     width="1000" height="600" alt="attributes list" />

In [15]:
titanic_ffill = titanic.ffill() 
titanic_ffill

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  19.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   C85        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500  C123        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   C50        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500  C148        Q  

[891 rows x 12 columns]

### *Optional:* Q13: Exame the previous new DataFrame resulted from forward filling method, and see if there are still missing (NA) values
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_ffill_count.png" 
     width="800" height="400" alt="attributes list" />

In [16]:
titanic_ffill.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          1
Embarked       0
dtype: int64

### Q14: Port of Embarkation is abbreviated noted in the Embarked column with either 'C','Q',or 'S' if the value is not missing. Please write an execution statement to verify the value categorical contents in column Embarked
<pre>    
Total unique values from Embarked column:
S C Q nan 
</pre>

In [17]:
titanic['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

### Q15: Use the dictionary mapping to create a new column named PortName, to accommodate the full names of the Embarked.  
<pre>
port_to_fullname = {'C': 'Cherbourg', 'Q': 'Queenstown', 'S':'Southampton'}
</pre>
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_embark_map.png" 
     width="1000" height="600" alt="attributes list" />

In [18]:
port_to_fullname = {'C': 'Cherbourg', 'Q': 'Queenstown', 'S':'Southampton'}
titanic['PortName'] = titanic['Embarked'].map(port_to_fullname)
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked     PortName  
0        0         A/5 21171   7.2500   NaN        S  Southampton  
1        0          PC 17599  71.2833   C85        C    Cherbourg  
2        0  STON/O2. 3101282   7.9250   NaN        S  Southampton  
3        0            113803  53.1000  C123        S  Southampton  
4        0            373450   8.0500   NaN        S  Southampton  
..     ...               ...      ...   ...      ...          ...  
886      0            211536  13.0000   NaN        S  Southampton  
887      0            112053  30.0000   B42        S  Southampton  
888      2        W./C. 6607  23.4500   NaN        S  Southampton  
889      0            111369  30.0000  C148        C    Cherbourg  
890      0            370376   7.7500   NaN        Q   Queenstown  

[891 rows x 13 columns]

### Q16: Print out the index and PassengerId of the first five entries from the original Titanic dataset.
<pre>
0    1
1    2
2    3
3    4
4    5
Name: PassengerId, dtype: int64
</pre>

In [19]:
titanic['PassengerId'][:5]

0    1
1    2
2    3
3    4
4    5
Name: PassengerId, dtype: int64

### *Optional:* Q17: Rename row indices to be aligned with PassengerId in place.
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_map_index.png" 
     width="1000" height="600" alt="attributes list" />

In [20]:
titanic.set_index('PassengerId', drop=False, inplace=True)
titanic

PassengerId  Survived  Pclass  \
PassengerId                                  
1                      1         0       3   
2                      2         1       1   
3                      3         1       3   
4                      4         1       1   
5                      5         0       3   
...                  ...       ...     ...   
887                  887         0       2   
888                  888         1       1   
889                  889         0       3   
890                  890         1       1   
891                  891         0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   
...            ...    ...               ...      ...   ...      ...   
887              0      0            211536  13.0000   NaN        S   
888              0      0            112053  30.0000   B42        S   
889              1      2        W./C. 6607  23.4500   NaN        S   
890              0      0            111369  30.0000  C148        C   
891              0      0            370376   7.7500   NaN        Q   

                PortName  
PassengerId               
1            Southampton  
2              Cherbourg  
3            Southampton  
4            Southampton  
5            Southampton  
...                  ...  
887          Southampton  
888          Southampton  
889          Southampton  
890            Cherbourg  
891           Queenstown  

[891 rows x 13 columns]

### Q18: Display the age distributions of all passengers in the dataset Using the age bins in the list.
<pre>
ages = [12, 18, 25, 35, 60, 80]
</pre>
<pre>
(25, 35]    196
(35, 60]    195
(18, 25]    162
(12, 18]     70
(60, 80]     22
Name: Age, dtype: int64
</pre>

In [21]:
ages = [12, 18, 25, 35, 60, 80]

In [22]:
pd.cut(titanic['Age'], bins=ages).value_counts()

Age
(25, 35]    196
(35, 60]    195
(18, 25]    162
(12, 18]     70
(60, 80]     22
Name: count, dtype: int64

### *Optional*: Q19: Display the age ranks of all passengers in the dataset using the categories where the whole age range of all passengers is divided into equal-length.
<pre>
group_names = ['Youth', 'YoungAdult', 'MiddleAged', 'Senior']
</pre>
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_age_bin.png" 
     width="600" height="900" alt="attributes list" />

In [23]:
group_names = ['Youth', 'YoungAdult', 'MiddleAged', 'Senior']
bins = [18, 25, 35, 60, 100]

In [24]:
min_age = titanic['Age'].min() 
max_age = titanic['Age'].max()

bin_width = (max_age - min_age) / 4
bins = [min_age, min_age + bin_width, min_age + 2*bin_width, min_age + 3*bin_width, max_age]

pd.cut(titanic['Age'], bins=bins, labels=group_names).value_counts()

Age
YoungAdult    385
Youth         178
MiddleAged    128
Senior         22
Name: count, dtype: int64

### Q20: Display the age categories in quartiles of all passengers in the Titanic dataset. 
<pre>
(20.125, 28.0]     183
(0.419, 20.125]    179
(38.0, 80.0]       177
(28.0, 38.0]       175
Name: Age, dtype: int64
</pre>

In [25]:
pd.qcut(titanic['Age'], q=4, labels=group_names).value_counts()

Age
YoungAdult    183
Youth         179
Senior        177
MiddleAged    175
Name: count, dtype: int64

### Q21: Detect Fare outliners, assuming there shouldn't be free ($0) ticket. 
There are 15 zero dolloar tickets.

In [26]:
titanic[titanic['Fare'] == 0]

PassengerId  Survived  Pclass                              Name  \
PassengerId                                                                    
180                  180         0       3               Leonard, Mr. Lionel   
264                  264         0       1             Harrison, Mr. William   
272                  272         1       3      Tornquist, Mr. William Henry   
278                  278         0       2       Parkes, Mr. Francis "Frank"   
303                  303         0       3   Johnson, Mr. William Cahoone Jr   
414                  414         0       2    Cunningham, Mr. Alfred Fleming   
467                  467         0       2             Campbell, Mr. William   
482                  482         0       2  Frost, Mr. Anthony Wood "Archie"   
598                  598         0       3               Johnson, Mr. Alfred   
634                  634         0       1     Parr, Mr. William Henry Marsh   
675                  675         0       2        Watson, Mr. Ennis Hastings   
733                  733         0       2              Knight, Mr. Robert J   
807                  807         0       1            Andrews, Mr. Thomas Jr   
816                  816         0       1                  Fry, Mr. Richard   
823                  823         0       1   Reuchlin, Jonkheer. John George   

              Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  \
PassengerId                                                          
180          male  36.0      0      0    LINE   0.0   NaN        S   
264          male  40.0      0      0  112059   0.0   B94        S   
272          male  25.0      0      0    LINE   0.0   NaN        S   
278          male   NaN      0      0  239853   0.0   NaN        S   
303          male  19.0      0      0    LINE   0.0   NaN        S   
414          male   NaN      0      0  239853   0.0   NaN        S   
467          male   NaN      0      0  239853   0.0   NaN        S   
482          male   NaN      0      0  239854   0.0   NaN        S   
598          male  49.0      0      0    LINE   0.0   NaN        S   
634          male   NaN      0      0  112052   0.0   NaN        S   
675          male   NaN      0      0  239856   0.0   NaN        S   
733          male   NaN      0      0  239855   0.0   NaN        S   
807          male  39.0      0      0  112050   0.0   A36        S   
816          male   NaN      0      0  112058   0.0  B102        S   
823          male  38.0      0      0   19972   0.0   NaN        S   

                PortName  
PassengerId               
180          Southampton  
264          Southampton  
272          Southampton  
278          Southampton  
303          Southampton  
414          Southampton  
467          Southampton  
482          Southampton  
598          Southampton  
634          Southampton  
675          Southampton  
733          Southampton  
807          Southampton  
816          Southampton  
823          Southampton

### Q22: Convert categorical data on Embarked column into a dummy mattrix with C, Q, S as indicators
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_embark_dummy.png" 
     width="150" height="300" alt="attributes list" />

In [27]:
pd.get_dummies(titanic['Embarked']).replace({True: 1, False: 0})

/var/folders/t6/jzgwrdks2vq53w20jlljrk840000gn/T/ipykernel_44612/192178074.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  pd.get_dummies(titanic['Embarked']).replace({True: 1, False: 0})


C  Q  S
PassengerId         
1            0  0  1
2            1  0  0
3            0  0  1
4            0  0  1
5            0  0  1
...         .. .. ..
887          0  0  1
888          0  0  1
889          0  0  1
890          1  0  0
891          0  1  0

[891 rows x 3 columns]

### *Optional:* Q23: Convert age distributions of all passengers in the Titanic dataset into a dummy mattrix  with the age bins in the list.
<pre>
ages = [12, 18, 25, 35, 60, 80]
</pre>
<img src="http://mathqs.mcdaniel.edu/ANA522/mod03/img/titanic_age_dummy.png" 
     width="300" height="500" alt="attributes list" />

In [28]:
ages = [12, 18, 25, 35, 60, 80]
pd.get_dummies(pd.cut(titanic['Age'], bins=ages)).replace({True: 1, False: 0})

/var/folders/t6/jzgwrdks2vq53w20jlljrk840000gn/T/ipykernel_44612/3114323977.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  pd.get_dummies(pd.cut(titanic['Age'], bins=ages)).replace({True: 1, False: 0})


(12, 18]  (18, 25]  (25, 35]  (35, 60]  (60, 80]
PassengerId                                                  
1                   0         1         0         0         0
2                   0         0         0         1         0
3                   0         0         1         0         0
4                   0         0         1         0         0
5                   0         0         1         0         0
...               ...       ...       ...       ...       ...
887                 0         0         1         0         0
888                 0         1         0         0         0
889                 0         0         0         0         0
890                 0         0         1         0         0
891                 0         0         1         0         0

[891 rows x 5 columns]